src/app/clusters/window-covering-server/window-covering-server.cpp

```diff
@@ -0,0 +1,650 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+/****************************************************************************
+ * @file
+ * @brief Routines for the Window Covering Server cluster
+ *******************************************************************************
+ ******************************************************************************/
+
+#include "window-covering-server.h"
+
+#include <app/util/af.h>
+
+#include <app/common/gen/attribute-id.h>
+#include <app/common/gen/cluster-id.h>
+#include <app/common/gen/command-id.h>
+
+#include <app/Command.h>
+#include <app/reporting/reporting.h>
+#include <app/util/af-event.h>
+#include <app/util/af-types.h>
+#include <app/util/attribute-storage.h>
+
+#ifdef EMBER_AF_PLUGIN_SCENES
+#include <app/clusters/scenes/scenes.h>
+#endif // EMBER_AF_PLUGIN_SCENES
+
+using namespace chip;
+
+#define CHECK_BOUNDS_INVALID(MIN, VAL, MAX) ((VAL < MIN) || (VAL > MAX))
+#define CHECK_BOUNDS_VALID(MIN, VAL, MAX) (!CHECK_BOUNDS_INVALID(MIN, VAL, MAX))
+
+#define WC_PERCENTAGE_COEF 100
+#define WC_PERCENT100THS_MAX 10000
+
+static EmberAfStatus writeAttribute(chip::EndpointId ep, chip::AttributeId attributeID, uint8_t * dataPtr,
+                                    EmberAfAttributeType dataType)
+{
+    if (!dataPtr)
+        return EMBER_ZCL_STATUS_INVALID_FIELD;
+
+    EmberAfStatus status =
+        emberAfWriteAttribute(ep, ZCL_WINDOW_COVERING_CLUSTER_ID, attributeID, CLUSTER_MASK_SERVER, dataPtr, dataType);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        emberAfWindowCoveringClusterPrint("Err: WC Writing Attribute failed, ep:%u, attr:%04x, err:%04x", ep, attributeID, status);
+    }
+    return status;
+}
+
+static EmberAfStatus readAttribute(chip::EndpointId ep, chip::AttributeId attributeID, uint8_t * dataPtr, uint16_t readLength)
+{
+    if (!dataPtr)
+        return EMBER_ZCL_STATUS_INVALID_FIELD;
+
+    EmberAfStatus status =
+        emberAfReadAttribute(ep, ZCL_WINDOW_COVERING_CLUSTER_ID, attributeID, CLUSTER_MASK_SERVER, dataPtr, readLength, NULL);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        emberAfWindowCoveringClusterPrint("Err: WC Reading Attribute failed, ep:%u, attr:%04x, err:%04x", ep, attributeID, status);
+    }
+
+    return status;
+}
+
+//------------------------------------------------------------------------------
+// WindowCover
+//------------------------------------------------------------------------------
+
+WindowCover WindowCover::sInstance;
+
+WindowCover & WindowCover::Instance()
+{
+    return sInstance;
+}
+
+WindowCover::WindowCover() {}
+
+void WindowCover::TypeSet(EmberAfWcType type)
+{
+    writeAttribute(mEndPoint, ZCL_WC_TYPE_ATTRIBUTE_ID, (uint8_t *) &type, ZCL_ENUM8_ATTRIBUTE_TYPE);
```

Comment:
All the casts below use C-style casts of variables and corresponding sizeof.

I would feel better if we could:
  - use `typename underlying_type<T>::type` to define the cpp type for the enum for the purposes of variable declaration and sizeof during reading/writing
  - use C++-style casts instead of C-style

Reply:
C-style casts replaced with C++-style.